v1/handler.js

```diff
@@ -0,0 +1,175 @@
+    var self = this;
+    var id = reqFrame.header.id;
+    var name = reqFrame.arg1.toString();
+
+    if (name === 'TChannel identify') {
+        self.handleInitRequest(reqFrame);
+        return;
+    }
+
+    if (self.conn.remoteName === null) {
+        self.conn.resetAll(new Error('call request before init request')); // TODO typed error
+        return;
+    }
+
+    var handler = self.conn.channel.getEndpointHandler(name);
+    self.conn.runInOp(handler, {
+        id: id,
+        arg1: reqFrame.arg1,
+        arg2: reqFrame.arg2,
+        arg3: reqFrame.arg3,
+    }, function buildResponseFrame(err, res1, res2) {
+        var resFrame = new v1.Frame();
+        resFrame.header.id = id;
+        resFrame.header.seq = 0;
+        if (err) {
+            // TODO should the error response contain a head ?
+            // Is there any value in sending meta data along with
+            // the error.
+            resFrame.set(isError(err) ? err.message : err, null, null);
+            resFrame.header.type = v1.Types.resError;
+        } else {
+            resFrame.set(name, res1, res2);
+            resFrame.header.type = v1.Types.resCompleteMessage;
+        }
+        return resFrame;
+    });
+};
+
+TChannelHandler.prototype.handleCallResponse = function handleCallResponse(resFrame) {
+    var self = this;
+
+    if (String(resFrame.arg1) === 'TChannel identify') {
+        self.handleInitResponse(resFrame);
+        return;
+    }
+
+    if (self.conn.remoteName === null) {
+        self.conn.resetAll(new Error('call response before init response')); // TODO typed error
+        return;
+    }
+
+    var id = resFrame.header.id;
+    var arg2 = resFrame.arg2;
+    var arg3 = resFrame.arg3;
+    self.conn.completeOutOp(id, null, arg2, arg3);
+};
+
+TChannelHandler.prototype.handleError = function handleError(errFrame) {
+    var self = this;
+    var id = errFrame.header.id;
+    var message = errFrame.arg1;
+    var err = new Error(message);
+    self.conn.completeOutOp(id, err, null, null);
+};
+
+TChannelHandler.prototype.sendInitRequest = function sendInitRequest() {
+    var self = this;
+    var reqFrame = new v1.Frame();
+    var id = self.conn.nextFrameId();
+    reqFrame.header.id = id;
+    reqFrame.header.seq = 0;
+    reqFrame.set('TChannel identify', self.conn.channel.hostPort, null);
+    reqFrame.header.type = v1.Types.reqCompleteMessage;
+    self.writeFrame(reqFrame);
+};
+
+TChannelHandler.prototype.sendInitResponse = function sendInitResponse(reqFrame) {
+    var self = this;
+    var id = reqFrame.header.id;
+    var arg1 = reqFrame.arg1;
+    var resFrame = new v1.Frame();
+    resFrame.header.id = id;
+    resFrame.header.seq = 0;
+    resFrame.set(arg1, self.conn.channel.hostPort, null);
+    resFrame.header.type = v1.Types.resCompleteMessage;
+    self.writeFrame(resFrame);
+};
+
+function isError(obj) {
+    return typeof obj === 'object' && (
+        Object.prototype.toString.call(obj) === '[object Error]' ||
```

Comment:
Sorry, that was terse.

typeof obj === 'object' is not the same as isObject(obj) for obj === null, which is the only case it's actually guarding against. Object.prototype.toString.call(null) and Object.prototype.toString.call(undefined) do not throw. If obj instanceof Error is true, then Object.prototype.toString.call(obj) is guaranteed to return '[object Error]', so it is superfluous, unless someone has replaced the global Error. The reason Object.prototype.toString.call(obj) is better than instanceof Error is because of Error from another context/realm.